### PR TITLE
Extensible DispatchHandler

### PR DESCRIPTION
**Pull request by arontsang**
This adds a plugin system that allows decorators to be added to the DispatchHandler as part of a pipeline.

I expect to see in future plugins such as (but not limited to):

- Windows Authentication NTLM/Negotiate/Kerberos
- Client side caching (https://github.com/nodejs/undici/issues/1146)
- Retry logic
- Telemetry/Logging

This should support https://github.com/nodejs/undici/issues/491

### CODE REVIEW

**Comment by ronag** [comment]
I don't really see what this brings other than maybe better ergonomics? The dispatcher interface is already quite extensible? Could this live outside as an npm package?

**Review by mcollina — CHANGES_REQUESTED** [review]
I see the need for better ergonomics for this, however I would not add any new globals for this. Could this be a feature of Agent instead?

**Comment by arontsang** [comment]
At the end of the day, this is simply a proposal.

I expect some back and forth before a final revision.

Personally, I'm not happy with the shape of the API, but I'm happy to take
suggestions.

Additionally, I'm considering switching to a DispatcherDecorator model,
with a AgentBuilder for setting the global dispatcher.

**Comment by arontsang** [comment]
Also, I'm having trouble coming up with an API that would not create a
breaking change on a raw Client, whilst also extracting out the
RedirectHandler.

**Comment by mcollina** [comment]
How would the DispatcherDecorator work?

**Comment by arontsang** [comment]
@mcollina Here is another commit showing what I mean https://github.com/arontsang/undici/commit/d54794f34480819103bf304476f7e26da4938a88?diff=unified

In effect we add to the constructor of Dispatcher `{ interceptors: Record<string, DispatchInterceptor[]> }`

Where `type DispatchInterceptor = Dispatcher[kDispatch] => Dispatcher[kDispatch]`
Where `type Dispatcher[kDispatch] = (options: Dispatcher.DispatchOptions, handler: Dispatcher.DispatchHandlers) => boolean`

We then reduce the array of `DispatchInterceptors` over `this[kDispatch]` to produce an intercepted dispatch method.

**Comment by arontsang** [comment]
I've just noticed that ProxyAgent could be replaced with a DispatchInterceptor quite trivially.

File: lib/dispatcher-base.js [hunk 1]
```diff
@@ -125,6 +130,15 @@ class DispatcherBase extends Dispatcher {
     })
   }
 
+  [kBuildInterceptedDispatch] () {
+    let dispatch = this[kDispatch].bind(this)
+    for (let i = this[kInterceptors].length - 1; i >= 0; i--) {
+      dispatch = this[kInterceptors][i](dispatch)
+    }
+
+    return dispatch
```
Comment (mcollina) [inline lib/dispatcher-base.js:139]: Can we cache this?

File: types/dispatcher.d.ts [hunk 2]
```diff
@@ -36,6 +37,59 @@ declare class Dispatcher extends EventEmitter {
...
+
+  rawListeners(eventName: 'connect'): ((origin: URL, targets: Array<Dispatcher>) => void)[]
+  rawListeners(eventName: 'disconnect'): ((origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void)[];
+  rawListeners(eventName: 'connectionError'): ((origin: URL, targets: Array<Dispatcher>, error: UndiciError) => void)[];
+  rawListeners(eventName: 'drain'): ((origin: URL) => void)[];
+
+  emit(eventName: 'connect', origin: URL, targets: Array<Dispatcher>): boolean;
+  emit(eventName: 'disconnect', origin: URL, targets: Array<Dispatcher>, error: UndiciError): boolean;
+  emit(eventName: 'connectionError', origin: URL, targets: Array<Dispatcher>, error: UndiciError): boolean;
+  emit(eventName: 'drain', origin: URL): boolean;
```
Comment (mcollina) [inline types/dispatcher.d.ts:92]: Can you add the tests for those?

File: lib/interceptor/redirect.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (mcollina) [inline lib/interceptor/redirect.js:1]: Add use strict to the top.

File: types/pool.d.ts [hunk 4]
```diff
@@ -18,5 +19,7 @@ declare namespace Pool {
     factory?(origin: URL, opts: object): Dispatcher;
     /** The max number of clients to create. `null` if no limit. Default `null`. */
     connections?: number | null;
+
+    interceptors?: { Pool?: DispatchInterceptor[] } & Client.Options["interceptors"]
```
Comment (mcollina) [inline types/pool.d.ts:23]: can you please add a test for these new types?

**Review by mcollina — COMMENTED** [review]
Good work! Could you run the benchmarks before/after? Could you document this new feature?

File: types/dispatcher.d.ts (hunk 2 above)
Comment (arontsang) [reply types/dispatcher.d.ts:92]: Sorry I am really having trouble with `tsd`.

I am much more comfortable writing Typescript than this mix. I am finding that the usage of `tsd` instead of writing the library in typescript is actually creating a lot of bugs in the `.d.ts` files (such as `DispatchHandler.onHeaders`).

Would you guys be open to converting the library to typescript instead?

File: types/pool.d.ts (hunk 4 above)
Comment (arontsang) [reply types/pool.d.ts:23]: I've added a few Jest tests to run some interceptors. Mainly in manipulating requests and responses.

File: types/dispatcher.d.ts (hunk 2 above)
Comment (mcollina) [reply types/dispatcher.d.ts:92]: We are ok with Undici being JavaScript, thanks for the offer.

File: lib/dispatcher-base.js (hunk 1 above)
Comment (arontsang) [reply lib/dispatcher-base.js:147]: Cached

File: types/dispatcher.d.ts (hunk 2 above)
Comment (arontsang) [reply types/dispatcher.d.ts:92]: @mcollina I am unable to write the tests for these events. I am not familiar enough with `tsd`. How would you like to proceed?

Comment (mcollina) [reply types/dispatcher.d.ts:92]: check out https://github.com/nodejs/undici/tree/main/test/types and tsd README. If you can write TypeScript, you could learn tsd quickly.

File: docs/api/DispatchInterceptor.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+### Basic Example 2
+
+Here is a contrived example of an interceptor stripping the headers from a response.
+
+```js
+'use strict'
+
+const clearHeadersInterceptor = dispatch => {
+  const DecoratorHandler = require('undici/lib/handler/decorator')
```
Comment (mcollina) [inline docs/api/DispatchInterceptor.md:45]: This is an internal library, let's not document it. If it needs to be exposed, please use the module entry point.

File: lib/dispatcher-base.js [hunk 6]
```diff
@@ -143,7 +148,26 @@ class DispatcherBase extends Dispatcher {
         throw new ClientClosedError()
       }
 
-      return this[kDispatch](opts, handler)
+      if (!this[kInterceptedDispatch]) {
+        function buildInterceptedDispatch () {
```
Comment (mcollina) [inline lib/dispatcher-base.js:152]: Could you move this to a top level function?

**Review by mcollina — COMMENTED** [review]
Good work!

**Comment by mcollina** [comment]
@ronag PTAL

File: docs/api/DispatchInterceptor.md (hunk 5 above)
Comment (arontsang) [reply docs/api/DispatchInterceptor.md:45]: So? Anything needed to be done here?

**Comment by arontsang** [comment]
> @ronag PTAL

@ronag Any chance you can take a look?

File: docs/api/DispatchInterceptor.md (hunk 5 above)
Comment (arontsang) [reply docs/api/DispatchInterceptor.md:45]: I've also noticed that http headers being of type ByteArray means that each extension is likely to be to decode utf8 once each...

**Comment by ronag** [comment]
I'll have. A look next week.

**Comment by arontsang** [comment]
Thanks

**Comment by ronag** [comment]
I think this can be achieved without modifying existing internals and I think I would maybe prefer that, i.e. implement a `compose/decorate/intercept` (or other name) function that implements a compose dispatcher that applies this logic. So it would essentially be possible to do the following without any modification of existing code:

```js
const insertHeaderInterceptor = dispatch => {
  return function InterceptedDispatch(opts, handler){
    opts.headers.push('Authorization', 'Bearer [Some token]')
    return dispatch(opts, handler)
  }
}

const client = undici.intercept(new Client('https://localhost:3000', [ insertHeaderInterceptor ])
```

**Comment by ronag** [comment]
Or maybe:

```js
const client = undici.compose([
  () => new Client('https://localhost:3000', 
  dispatch => function insertHeader(opts, handler){
    opts.headers.push('Authorization', 'Bearer [Some token]')
    return dispatch(opts, handler)
  }
])
```

**Comment by ronag** [comment]
```js
undici.request(url, [
   decorate: [insertHeaders({ foo: 'bar' }, redirect({ maxRetries: 3 )]
])
```

**Comment by ronag** [comment]
This might need a few iterations and since it doesn't require any internal modifications we should consider whether or not we should first have it as an external package. Not saying we should. Just to consider.

**Comment by arontsang** [comment]
> ```js
> undici.request(url, {
>    decorate: [insertHeaders({ foo: 'bar' }, redirect({ maxRetries: 3 )]
> })
> ```

This would not work for my use case. I NEED to apply the decorator AT the client level of an Agent.

The decorator needs to be scoped to a single TCP connection to be able to implement NTLM authentication.

**Comment by arontsang** [comment]
> Or maybe:
> 
> ```js
> const client = undici.compose(
>   () => new Client('https://localhost:3000', 
>   dispatch => function insertHeader(opts, handler){
>     opts.headers.push('Authorization', 'Bearer [Some token]')
>     return dispatch(opts, handler)
>   }
> )
> ```

This works, but it would require an inordinate amount of boiler plate to bring up a full Agent with decorators at the right levels, Agent, Pool, Client etc...

**Comment by ronag** [comment]
Good point. Please give me a little time to mull and experiment with this.

File: docs/api/DispatchInterceptor.md (hunk 5 above)
Comment (mcollina) [reply docs/api/DispatchInterceptor.md:45]: Go ahead and expose it in this PR.

> I've also noticed that http headers being of type ByteArray means that each extension is likely to be to decode utf8 once each...

What do you mean? Can you make an example?

**Comment by arontsang** [comment]
> Good point. Please give me a little time to mull and experiment with this.

Any thoughts?

**Comment by ronag** [comment]
Sorry for the delay. This is quite tricky and not something that's easy to land. It's good work but this kind major API change is something that we can't change back and forth and needs thought.

I will spend more time on this before I go on vacation next week.

**Comment by arontsang** [comment]
Well, if you are already thinking about it. I would like to change the
interface for the DispatchHandler to return headers as string[] rather than
ByteArray[].

This would help reduce the number of Utf8 decodes.

File: lib/dispatcher-base.js [hunk 7]
```diff
@@ -125,6 +130,24 @@ class DispatcherBase extends Dispatcher {
...
+  [kInterceptedDispatch] (opts, handler) {
+    if (!this[kInterceptors] || this[kInterceptors].length === 0) {
+      this[kInterceptedDispatch] = this[kDispatch]
+      return this[kDispatch](opts, handler)
+    }
+
+    let dispatch = this[kDispatch].bind(this)
+    for (let i = this[kInterceptors].length - 1; i >= 0; i--) {
+      const interceptor = this[kInterceptors][i]
+      if (typeof interceptor !== 'function') {
```
Comment (kibertoad) [inline lib/dispatcher-base.js:142]: Can we validate this once on addition and not on every dispatch?

File: lib/handler/decorator.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+'use strict'
+
+module.exports = class DecoratorHandler {
```
Comment (kibertoad) [inline lib/handler/decorator.js:3]: file name is inconsistent with class name

File: lib/interceptor/redirect.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+'use strict'
+
+const RedirectHandler = require('../handler/redirect')
+
+function RedirectInterceptor ({ maxRedirections: defaultMaxRedirections }) {
```
Comment (kibertoad) [inline lib/interceptor/redirect.js:5]: this has class naming but is not a class. maybe it should be renamed to indicate that this is a factory function for a partially resolved function?

Comment (kibertoad) [inline lib/interceptor/redirect.js:5]: this kind of renaming is fairly confusing, do we need it?

File: test/jest/interceptor.test.js [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    const origin = new URL(`http://localhost:${server.address().port}`)
+    await Promise.all([
+      request(origin, { dispatcher: agent }),
+      request(origin, { dispatcher: agent })
+    ])
+
+    // Assert that the requests are run on different interceptors (different Clients)
+    const requestCounts = interceptors.map(x => x.requestCount)
+    expect(requestCounts).toEqual([1, 1])
+  } finally {
```
Comment (kibertoad) [inline test/jest/interceptor.test.js:36]: why? can't this be handled in jest `after` block?

File: types/client.d.ts [hunk 11]
```diff
@@ -41,6 +42,8 @@ declare namespace Client {
     tls?: TlsOptions | null;
     /** */
     maxRequestsPerClient?: number;
+
+    interceptors?: {Client: DispatchInterceptor[] | undefined}
```
Comment (kibertoad) [inline types/client.d.ts:46]: do you need to mention `undefined` explicitly if it's an optional parameter already?

File: types/dispatcher.d.ts [hunk 12]
```diff
@@ -36,6 +37,59 @@ declare class Dispatcher extends EventEmitter {
   destroy(err: Error | null): Promise<void>;
   destroy(callback: () => void): void;
   destroy(err: Error | null, callback: () => void): void;
+
+  on(eventName: 'connect', callback: (origin: URL, targets: Array<Dispatcher>) => void): this;
```
Comment (kibertoad) [inline types/dispatcher.d.ts:41]: probably readonly arrays should suffice in all these cases?

File: types/dispatcher.d.ts [hunk 13]
```diff
@@ -145,7 +199,7 @@ declare namespace Dispatcher {
     /** Invoked when request is upgraded either due to a `Upgrade` header or `CONNECT` method. */
     onUpgrade?(statusCode: number, headers: string[] | null, socket: Duplex): void;
     /** Invoked when statusCode and headers have been received. May be invoked multiple times due to 1xx informational headers. */
-    onHeaders?(statusCode: number, headers: string[] | null, resume: () => void): boolean;
+    onHeaders?(statusCode: number, headers: Buffer[] | null, resume: () => void): boolean;
```
Comment (kibertoad) [inline types/dispatcher.d.ts:202]: why this change?

File: types/client.d.ts [hunk 14]
```diff
@@ -1,8 +1,9 @@
 import { URL } from 'url'
 import { TlsOptions } from 'tls'
 import Dispatcher = require('./dispatcher')
-import { DispatchOptions, RequestOptions } from './dispatcher'
-import buildConnector = require('./connector')
+import {DispatchInterceptor} from './dispatcher'
+import buildConnector, {connector} from "./connector";
+import {Socket} from "net"
```
Comment (kibertoad) [inline types/client.d.ts:6]: where is this used?

**Comment by kibertoad** [comment]
Is there something I could help with that would advance this PR further?

File: types/dispatcher.d.ts (hunk 13 above)
Comment (arontsang) [reply types/dispatcher.d.ts:202]: Because the typescript is wrong. I am merely correcting the type definition file.

To be honest, this is why I hate javascript libraries, nobody knows what the correct types are anywhere.

File: test/jest/interceptor.test.js (hunk 10 above)
Comment (arontsang) [reply test/jest/interceptor.test.js:36]: Because I am not a javascript programmer and I don't use jest. I had enough trouble getting this work...

**Review by ronag — COMMENTED** [review]
@mcollina I'm not convinced by the direction of this at the moment but I'm having difficulty getting time to focus on this and give it the thought that is required. I'm ok with merging this but I might propose semver major changes in the future. I'm not sure what the right thing to do here is. I don't want to be blocking here for unreasonably long time either.

**Comment by mcollina** [comment]
Let's go ahead and land. The code seems reasonable to me so far, minus all the comments.
I'm 100% ok with semver-major changes on this one.

Maybe we could classify interceptors as experimental?

Could you address them @arontsang? Sorry for the long wait.

**Comment by kibertoad** [comment]
@arontsang if you need any help, can you give me write permissions on your repo?

**Comment by arontsang** [comment]
> @arontsang if you need any help, can you give me write permissions on your repo?

Added

**Comment by kibertoad** [comment]
I'll finally have some free time next week and will try to resolve conflicts here, sorry for the delay

**Comment by ci010** [comment]
I think PR is really useful for extensibility.

I'm interested in implementing retry logic with it, and I did some POCs.
I found current design might have some gaps to implement retry.

Consider following retry handler (implemented by DecoratorHandler):
```ts

export class RetryHandler extends DecoratorHandler {
    private attempt = 0

    constructor(handler: Dispatcher.DispatchHandlers, private policy: RetryPolicy['retry'], private dispatcher: Dispatcher['dispatch'], private opts: Dispatcher.DispatchOptions) {
        super(handler)
    }

    onError(err: Error): void {
        const shouldRetry = this.policy(this.opts, ++this.attempt, err)
        if (typeof shouldRetry === 'boolean') {
            if (shouldRetry) {
                this.dispatcher(this.opts, this)
            } else {
                super.onError(err)
            }
        } else {
            shouldRetry.then((shouldRetry) => {
                if (shouldRetry) {
                    this.dispatcher(this.opts, this)
                } else {
                    super.onError(err)
                }
            })
        }
    }
}
```

Basically, it will re-dispatch the http request once the error happened and the error should be retried (by user logic).

This works totally fine if the error is thrown before the client receive the response `headers` and `statusCode`.
However, if the error is thrown during body receiving progress, the retry won't work, as the `Request`'s `callback` is set to null.
(Like body timeout error)

https://github.com/nodejs/undici/blob/main/lib/api/api-request.js#L90
```ts
onHeaders() {
...
    this.callback = null // set callback to null
    this.res = body
...
}
```

And that will make the further retries always failed:

https://github.com/nodejs/undici/blob/main/lib/api/api-request.js#L68
```ts
onConnect (abort, context) {
    if (!this.callback) {
      throw new RequestAbortedError() // no callback will throw abort error
    }

    this.abort = abort
    this.context = context
  }
```

The `stream` and `pipeline` have different behaviors. `stream` will clear the `callback` after the `response` stream finished. I didn't test it, but it might work. Maybe we can do the similar things in `request`.

I just wonder will this PR also deal this issue. Or it's better to be another PR.

**Comment by kibertoad** [comment]
@arontsang Can you provide any details on
```
Well, if you are already thinking about it. I would like to change the
interface for the DispatchHandler to return headers as string[] rather than
ByteArray[].

This would help reduce the number of Utf8 decodes.
```
? I can't find where in the code this is happening

**Comment by kibertoad** [comment]
@mcollina @ronag I've resolved conflicts and addressed all the outstanding comments, could you please rereview?

**Comment by kibertoad** [comment]
@mcollina @ronag All green and ready for rereview.

File: docs/api/DispatchInterceptor.md (hunk 5 above)
Comment (mcollina) [reply docs/api/DispatchInterceptor.md:45]: @kibertoad this needs fixing, we should not be deep-requiring.

File: docs/api/Pool.md [hunk 15]
```diff
@@ -19,6 +19,7 @@ Extends: [`ClientOptions`](Client.md#parameter-clientoptions)
 
 * **factory** `(origin: URL, opts: Object) => Dispatcher` - Default: `(origin, opts) => new Client(origin, opts)`
 * **connections** `number | null` (optional) - Default: `null` - The number of `Client` instances to create. When set to `null`, the `Pool` instance will create an unlimited amount of `Client` instances.
+* **interceptors.Pool** `Array<DispatchInterceptor>` - Default: `[]` - A list of interceptors that are applied to the dispatch method. Additional logic can be applied (such as, but not limited to: 302 status code handling, authentication, cookies, compression and caching).
```
Comment (mcollina) [inline docs/api/Pool.md:22]: It's not clear this is `{ interceptors: { Pool: [] } }`

**Comment by kibertoad** [comment]
@mcollina Comments addressed!

File: lib/proxy-agent.js [hunk 16]
```diff
@@ -42,7 +62,7 @@ class ProxyAgent extends DispatcherBase {
 
     const connect = buildConnector({ ...opts.proxyTls })
     this[kConnectEndpoint] = buildConnector({ ...opts.requestTls })
-    this[kClient] = new Client({ origin: opts.origin, connect })
+    this[kClient] = new Agent({ origin: opts.origin, connect })
```
Comment (mcollina) [inline lib/proxy-agent.js:65]: Why this change?

**Review by mcollina — APPROVED** [review]
I would avoid introducing any more global settings, could you avoid them? Thanxk

File: lib/proxy-agent.js (hunk 16 above)
Comment (kibertoad) [reply lib/proxy-agent.js:65]:
```
const Client = require('./agent')
const Agent = require('./agent')
```

Naming was confusing, they were exactly same thing. Am I missing anything?

Comment (ronag) [reply lib/proxy-agent.js:65]: Seems more like the import was incorrect?

Comment (kibertoad) [reply lib/proxy-agent.js:65]: ok, I'll fix it then

**Comment by kibertoad** [comment]
> I would avoid introducing any more global settings, could you avoid them? Thanxk

@mcollina Do you mean new options for Client/Agent/Pool? Where would you place them instead?

File: lib/proxy-agent.js (hunk 16 above)
Comment (kibertoad) [reply lib/proxy-agent.js]: done

File: docs/api/Pool.md (hunk 15 above)
Comment (kibertoad) [reply docs/api/Pool.md]: fixed in all docs

**Review by mcollina — APPROVED** [review]
lgtm, that text was a leftover from a previous half-botched review

**Review by ronag — APPROVED** [review]

**Comment by kibertoad** [comment]
@mcollina @ronag What's left before this can be merged?
Also it would be great to release a new version sometime after the merge, would allow our team to start working on actual Undici interceptors :D

**Comment by wong2** [comment]
Does this support async logic in interceptors? For example, read credentials from the disk in a request interceptor and then add them to request headers.

**Comment by wong2** [comment]
Also I find it annoying when extending request headers in a request interceptor, the `opts.headers` type is vague, according to typescript declaration they could be `IncomingHttpHeaders | string[] | null`, which makes it hard when extending...

**Comment by arontsang** [comment]
I would need help with some of this.

I am a dotnet Dev by trade, so a lot of it isn't my wheelhouse, for
example, I don't know the Jest library inside and out...

Would it be better if we put this into a branch?

**Comment by arontsang** [comment]
It can, but, that would get very complicated when we get to the point of
having half a dozen of these things
